cirq-core/cirq/transformers/routing/route_circuit_cqc.py

```diff
@@ -0,0 +1,445 @@
+        two_qubit_ops: List[List['cirq.Operation']],
+        idx: int,
+        max_search_radius: int,
+    ) -> Optional[Tuple[QidPair, ...]]:
+        """Computes cost function with pairs of candidate swaps that act on disjoint qubits."""
+        pair_sigma = disjoint_nc2_combinations(cls._initial_candidate_swaps(mm, two_qubit_ops[idx]))
+        return cls._iterate_swaps(mm, two_qubit_ops, idx, max_search_radius, pair_sigma)
+
+    @classmethod
+    def _iterate_single_swap_strategy(
+        cls,
+        mm: mapping_manager.MappingManager,
+        two_qubit_ops: List[List['cirq.Operation']],
+        idx: int,
+        max_search_radius: int,
+    ) -> Optional[Tuple[QidPair, ...]]:
+        """Computes cost function with list of single candidate swaps."""
+        sigma: List[Tuple[QidPair, ...]] = [
+            (swap,) for swap in cls._initial_candidate_swaps(mm, two_qubit_ops[idx])
+        ]
+        return cls._iterate_swaps(mm, two_qubit_ops, idx, max_search_radius, sigma)
+
+    @classmethod
+    def _iterate_swaps(
+        cls,
+        mm: mapping_manager.MappingManager,
+        two_qubit_ops: List[List['cirq.Operation']],
+        idx: int,
+        max_search_radius: int,
+        sigma: List[Tuple[QidPair, ...]],
+    ) -> Optional[Tuple[QidPair, ...]]:
+        """Computes cost function of a list of n-tuple candidate swaps."""
+        for s in range(idx, min(max_search_radius + idx, len(two_qubit_ops))):
+            if len(sigma) <= 1:
+                break
+            sigma = cls._get_next_swaps_with_min_cost(mm, sigma, two_qubit_ops[s])
+
+        return (
+            None if len(sigma) > 1 and idx + max_search_radius <= len(two_qubit_ops) else sigma[0]
+        )
+
+    @classmethod
+    def _initial_candidate_swaps(
+        cls, mm: mapping_manager.MappingManager, two_qubit_ops: List['cirq.Operation']
+    ) -> List[QidPair]:
+        """Finds all feasible SWAPs between qubits involved in 2-qubit operations."""
+        physical_qubits = (mm.map[op.qubits[i]] for op in two_qubit_ops for i in range(2))
+        physical_swaps = mm.induced_subgraph.edges(nbunch=physical_qubits)
+        return [(mm.inverse_map[q1], mm.inverse_map[q2]) for q1, q2 in physical_swaps]
+
+    @classmethod
+    def _get_next_swaps_with_min_cost(
+        cls,
+        mm: mapping_manager.MappingManager,
+        candidate_swaps: List[Tuple[QidPair, ...]],
+        two_qubit_ops: List['cirq.Operation'],
+    ) -> List[Tuple[QidPair, ...]]:
+        """Iterates the heuristic function.
+
+        Given a list of candidate swaps find a subset that leads to a minimal longest shortest path
+        between any paired qubits in the curernt timestep.
+        """
+        costs = {}
+        for swaps in candidate_swaps:
+            costs[swaps] = cls._cost(mm, swaps, two_qubit_ops)
+        _, min_cost = min(costs.items(), key=lambda x: x[1])
+        return [swaps for swaps, cost in costs.items() if cost == min_cost]
```

Comment:
We can consider copying this in the `_iterate_swaps` / `_choose_optimal_swap` method above since that's the only call site which calls this method and the implementation is fairly straightforward. But I don't have very strong opinions here.